feat: allow customization of tree view label / sorting to preserve old tree view behavior

hikchoi opened the pull request:
# feat: allow customization of tree view label / sorting to preserve old tree view behavior

This PR:
- adds a new config namespace `workspace.views` that will hold configurations related to dendron contributed views
- adds a config namespace `treeView` under `workspace.views`, with the key `treeItemLabelType`
- allow customization of the tree item label in tree view

- This change is added to preserve the old labeling behavior of tree view V1 for users who are already used to the tree view labeling notes with their file names.
- However, since the change to labeling was done because of the confusion introduced by the way we sort them, the sorting behavior is also tied to the labeling behavior.
  - If the items are labeled with the note's title, it will also be sorted by the title.
  - If the items are labeled with the note's fname, it will also be sorted by the fname.

## Code

### Basics

- [x] code should follow [Code Conventions](https://wiki.dendron.so/notes/773e0b5a-510f-4c21-acf4-2d1ab3ed741e.html)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](https://wiki.dendron.so/notes/pMS27sHxbWeKMoPRrWEzs.html).
- [x] sticking to existing conventions instead of creating new ones 
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended

- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [~] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed 
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)



## Instrumentation

### Basics

- [~] if you are adding analytics related changes, make sure you [Document](https://wiki.dendron.so/notes/8ThPaB9iXXm2Szk3C9kFt.html) changes in airtable

### Extended

- [~] can we track the performance of this change to know if it is _successful_? 
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

## 



## Tests

### Basics

- [x] [Write Tests](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html) 
- [x] [Confirm existing tests pass](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html)
- [x] [Confirm manual testing](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [x] If your tests changes an existing snapshot, snapshots have been [updated](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html)

### Extended

- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](https://wiki.dendron.so/notes/dtMsF12SF2SUhLN10sYe2.html)

- CSS
  - [~] display is correct for following dimensions
    - [~] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [~] lg: screen ≥ 992px
    - [~] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [~] display is correct for following browsers (across the various dimensions)
    - [~] safari
    - [~] firefox
    - [~] chrome



## Docs

### Basics

- [x] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

## 


### Basics

- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](https://wiki.dendron.so/notes/3489b652-cd0e-4ac8-a734-08094dc043eb.html) or [Packages](https://wiki.dendron.so/notes/32cdd4aa-d9f6-4582-8d0c-07f64a00299b.html)

## 



## Close the Loop

### Basics

### Extended

- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](https://wiki.dendron.so/notes/iZ8vpfY0n1E3Nq3IC1Y7X.html)
  - eg. [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

hikchoi commented:
DOCS: https://github.com/dendronhq/dendron-site/pull/486

jonathanyeung commented:
Instead of having this as a setting in `dendron.yml`, how about we add a button to the tree view that allows users to change the options there of displaying title vs fname?  Maybe we can either add a dedicated button or use the ellipses '...' button with menu (for example of what i'm referring to as ellipses, see VS's built-in tree view control for 'Source Control')

We can then store the setting value into our metdata svc so that it persists for the user.

I feel like this will make the setting much more discoverable and make it easier to toggle. 

Sorry, I know this is quite a bit different than what you've coded up - lemme know what you think.

hikchoi commented:
> Instead of having this as a setting in `dendron.yml`, how about we add a button to the tree view that allows users to change the options there of displaying title vs fname? Maybe we can either add a dedicated button or use the ellipses '...' button with menu (for example of what i'm referring to as ellipses, see VS's built-in tree view control for 'Source Control')
> 
> We can then store the setting value into our metdata svc so that it persists for the user.
> 
> I feel like this will make the setting much more discoverable and make it easier to toggle.
> 
> Sorry, I know this is quite a bit different than what you've coded up - lemme know what you think.

@jonathanyeung 

Yeah I briefly thought about that route as well. Do you think having this setting persist in metadata svc vs other settings controlled in `dendron.yml` won't create any confusion? 

Where do we draw the line of "this setting stays in `dendron.yml`" and "this setting is persisted in the metadata"? 
I didn't have a good answer to this question so I went with the usual route of adding a new config in `dendron.yml`.

Happy to discuss and change the implementation accordingly.

hikchoi commented:
As per @kevinslin's comment in the daily async, I will keep the current implementation of storing the setting in `dendron.yml`.

For the sake of iteration, I will:
- [ ] add a toggle button that will flip between different label/sort behavior
- [ ] have it persist only for the current session
- [ ] display a toast on toggle that the default could be set in `dendron.yml`

hikchoi commented:
update:

I still need to add some tests for the newly added features listed below, but this PR is ready for review otherwise.

- toggle button for label type added
- when toggled, a toast will appear that suggests setting default value
  - user will have an option to set the current label type as default in `dendron.yml` (as a button)
    - clicking will back up `dendron.yml`, and then change `workspace.view.treeView.treeItemLabelType` to the current one.
  - user will also have an option to dismiss this toast forever (as a button)
- added a _expand all_ button behind `DendronContext.DEV_MODE`.
  - this works fine, but it's a bit slow. 
    - Unfortunately I don't think the VSCode team has plans to add this feature to the tree view API yet, so we have to loop through expandable items and call `.reveal(...)`.
    - With the depth limited to 3 hierarchies, testing in org-workspace gives us around 7000ms to expand 600-ish tree items initially. After the tree is built, it takes around 2000ms. (if the tree is rebuilt, i.e. label type is changed, it will take 7000ms again)

toggle buttons:
![image](https://user-images.githubusercontent.com/1219789/166619870-024daa6a-36b9-415a-9b63-3688dbc74e7b.png)

toast message:
![image](https://user-images.githubusercontent.com/1219789/166619907-0c1ec8e9-2b98-4a9b-8dce-39f856ba37f1.png)

toast message after clicking Set to default:
![image](https://user-images.githubusercontent.com/1219789/166620100-e33ad018-532a-42d2-bd5e-83037b3c0675.png)

note: if you manually test this against `org-workspace`, these errors will pop up.
these are actual notes that have duplicate ids in our workspace.: 
![image](https://user-images.githubusercontent.com/1219789/166619971-beb8644e-d382-483e-a9b0-1099a4affb34.png)

hikchoi commented:
related to #2739

hikchoi commented:
Update: UI state is now persisted using the metadata store instead of config. As a result the toast message no longer prompts for setting the default value

jonathanyeung commented on packages/plugin-core/src/_extension.ts:
```diff
@@ -1255,6 +1258,54 @@ function _setupLanguageFeatures(context: vscode.ExtensionContext) {
+   * Leaving this for dev / debug purposes.
+   * Enablement is set to be DendronContext.DEV_MODE
+   *
+   * TODO: fix tree item register issue and flip the dev mode flag.
```
out of curiosity, the flakiness is that sometimes it throws that 'duplicate id registration' error?

hikchoi replied:
That is _mostly_ fixed. it still happens but very rarely it seems. I removed the cause for like 95% of the cases but I'm yet to figure out the rest.

Also not mentioned here, but we are at worst calling n (numNotes) different async callbacks to fetch the file so it's pretty slow. The main reason VSCode doesn't provide this as a standard API.

I would ideally want to change this to be tree item specific (possible, add it as an inline button for each item), or have it only expand visible tree items (not sure if possible). Haven't specced out the journey here yet.

I still would like to have this behind the dev flag even if we make it tree item specific though, as this is pretty useful for visually debugging the tree view.

jonathanyeung replied:
cool, thanks for the detailed explanation.

jonathanyeung reviewed (approved):
Looks great!

Tested locally, seems to work well.  Maybe a few nitpicks UI wise, but we can refine it in a later iteration
- I don't think it's necessary to toast everytime we click the button.  Since every line will at least change capitalization, there's quite a lot of visual feedback that the button click is doing something. 
- my pref would be to align with VS Code's panels in having the sort order be a '...' context menu, but more of a nitpick.  (I'm not sure if they expose that control to extensions?)

hikchoi commented:
> I don't think it's necessary to toast everytime we click the button. Since every line will at least change capitalization, there's quite a lot of visual feedback that the button click is doing something.

I can fix this. would be a quick change.

> my pref would be to align with VS Code's panels in having the sort order be a '...' context menu, but more of a nitpick. (I'm not sure if they expose that control to extensions?)

Completely missed that we discussed about this. Yes this is available to extension authors. The contribution point is `view/title` without the `group` specification (the individual icons have `group: navigation`): https://code.visualstudio.com/api/extension-guides/tree-view#view-actions

I'll get this in for this PR. I don't like the idea of introducing a visible UI element and taking it away in short succession 😄

kevinslin reviewed (approved):
lgtm. feel free to merge when you feel is ready

hikchoi commented:
Update: moved the toggle button here
![image](https://user-images.githubusercontent.com/1219789/168163013-f9498fd6-8420-44ad-89f5-4b56d85a7279.png)

This will cycle through.

@jonathanyeung, fyi: I added `dendron-view` and other contribution points that are currently in `master` that are not set up to automatically generated with `gen:config`.

See https://github.com/dendronhq/dendron/pull/2858/commits/bcbe0768fc25a97f3bc4fff2eadc8995cf4a33d2
